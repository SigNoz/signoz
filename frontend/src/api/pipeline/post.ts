import axios from 'api';
import { ErrorResponseHandlerV2 } from 'api/ErrorResponseHandlerV2';
import { AxiosError } from 'axios';
import { ErrorV2Resp, SuccessResponseV2 } from 'types/api';
import { Pipeline } from 'types/api/pipeline/def';
import { Props } from 'types/api/pipeline/post';

const post = async (props: Props): Promise<SuccessResponseV2<Pipeline>> => {
	try {
		const response = await axios.post('/logs/pipelines', props.data);

		return {
			httpStatusCode: response.status,
			data: response.data.data,
		};
	} catch (error) {
		ErrorResponseHandlerV2(error as AxiosError<ErrorV2Resp>);
	}
};

export default post;
