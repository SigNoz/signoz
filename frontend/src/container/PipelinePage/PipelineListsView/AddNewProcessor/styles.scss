.processor-form-container {
	position: relative;
	width: 100%;

	display: flex;
	flex-wrap: wrap;
}

.processor-field-container {
	display: flex;
	flex-direction: row;
	align-items: flex-start;
	padding: 0rem;
	gap: 1rem;
	width: 100%;
}

.compact-processor-field-container {
	display: flex;
	flex-direction: row;
	align-items: flex-start;
	padding: 0rem;
	min-width: 40%;
	flex-grow: 1;
	margin-left: 2.5rem;
}
