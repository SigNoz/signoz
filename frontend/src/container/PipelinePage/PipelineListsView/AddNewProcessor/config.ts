import { FormInstance } from 'antd';
import { Rule, RuleRender } from 'antd/es/form';
import { NamePath } from 'antd/es/form/interface';
import { ProcessorData } from 'types/api/pipeline/def';

type ProcessorType = {
	key: string;
	value: string;
	label: string;
	title?: string;
	disabled?: boolean;
};

export const processorTypes: Array<ProcessorType> = [
	{ key: 'grok_parser', value: 'grok_parser', label: 'Grok' },
	{ key: 'regex_parser', value: 'regex_parser', label: 'Regex' },
	{ key: 'json_parser', value: 'json_parser', label: 'Json Parser' },
	{ key: 'trace_parser', value: 'trace_parser', label: 'Trace Parser' },
	{ key: 'time_parser', value: 'time_parser', label: 'Timestamp Parser' },
	{ key: 'severity_parser', value: 'severity_parser', label: 'Severity Parser' },
	{ key: 'add', value: 'add', label: 'Add' },
	{ key: 'remove', value: 'remove', label: 'Remove' },
	// { key: 'retain', value: 'retain', label: 'Retain' }, @Chintan - Commented as per Nitya's suggestion
	{ key: 'move', value: 'move', label: 'Move' },
	{ key: 'copy', value: 'copy', label: 'Copy' },
];

export const DEFAULT_PROCESSOR_TYPE = processorTypes[0].value;

export type ProcessorFieldOption = {
	label: string;
	value: string;
};

// TODO(Raj): Refactor Processor Form code after putting e2e UI tests in place.
export type ProcessorFormField = {
	id: number;
	fieldName: string;
	placeholder: string;
	name: string | NamePath;
	rules?: Array<Rule>;
	hidden?: boolean;
	initialValue?: boolean | string | Array<string>;
	dependencies?: Array<string | NamePath>;
	options?: Array<ProcessorFieldOption>;
	shouldRender?: (form: FormInstance) => boolean;
	onFormValuesChanged?: (
		changedValues: ProcessorData,
		form: FormInstance,
	) => void;

	// Should this field have its own row or should it
	// be packed with other compact fields.
	compact?: boolean;
};

const traceParserFieldValidator: RuleRender = (form) => ({
	validator: (): Promise<void> => {
		const parseFromValues = [
			['trace_id', 'parse_from'],
			['span_id', 'parse_from'],
			['trace_flags', 'parse_from'],
		].map((np) => form.getFieldValue(np));

		if (!parseFromValues.some((v) => v?.length > 0)) {
			return Promise.reject(
				new Error('At least one of the trace parser fields must be specified.'),
			);
		}

		return Promise.resolve();
	},
});

const commonFields = [
	{
		id: 3,
		fieldName: 'Parse From',
		placeholder: 'processor_parsefrom_placeholder',
		name: 'parse_from', // optional
		rules: [],
		initialValue: 'body',
	},
	{
		id: 4,
		fieldName: 'Parse To',
		placeholder: 'processor_parseto_placeholder',
		name: 'parse_to', // optional
		rules: [],
		initialValue: 'attributes',
	},
	{
		id: 5,
		fieldName: 'On Error',
		placeholder: 'processor_onerror_placeholder',
		name: 'on_error', // optional
		rules: [],
		initialValue: 'send',
	},
];

export const processorFields: { [key: string]: Array<ProcessorFormField> } = {
	grok_parser: [
		{
			id: 1,
			fieldName: 'Name of Grok Processor',
			placeholder: 'processor_name_placeholder',
			name: 'name',
		},
		{
			id: 2,
			fieldName: 'Pattern',
			placeholder: 'processor_pattern_placeholder',
			name: 'pattern',
		},
		...commonFields,
	],
	json_parser: [
		{
			id: 1,
			fieldName: 'Name of Json Parser Processor',
			placeholder: 'processor_name_placeholder',
			name: 'name',
		},
		{
			id: 2,
			fieldName: 'Parse From',
			placeholder: 'processor_parsefrom_placeholder',
			name: 'parse_from',
			initialValue: 'body',
		},
		{
			id: 3,
			fieldName: 'Parse To',
			placeholder: 'processor_parseto_placeholder',
			name: 'parse_to',
			initialValue: 'attributes',
		},
		{
			id: 4,
			fieldName: 'Enable Flattening',
			placeholder: '',
			name: 'enable_flattening',
			initialValue: false,
		},
	],
	regex_parser: [
		{
			id: 1,
			fieldName: 'Name of Regex Processor',
			placeholder: 'processor_name_placeholder',
			name: 'name',
		},
		{
			id: 2,
			fieldName: 'Define Regex',
			placeholder: 'processor_regex_placeholder',
			name: 'regex',
		},
		...commonFields,
	],
	add: [
		{
			id: 1,
			fieldName: 'Name of Add Processor',
			placeholder: 'processor_name_placeholder',
			name: 'name',
		},
		{
			id: 2,
			fieldName: 'Field',
			placeholder: 'processor_field_placeholder',
			name: 'field',
		},
		{
			id: 3,
			fieldName: 'Value',
			placeholder: 'processor_value_placeholder',
			name: 'value',
		},
	],
	remove: [
		{
			id: 1,
			fieldName: 'Name of Remove Processor',
			placeholder: 'processor_name_placeholder',
			name: 'name',
		},
		{
			id: 2,
			fieldName: 'Field',
			placeholder: 'processor_field_placeholder',
			name: 'field',
		},
	],
	trace_parser: [
		{
			id: 1,
			fieldName: 'Name of Trace Parser Processor',
			placeholder: 'processor_name_placeholder',
			name: 'name',
		},
		{
			id: 2,
			fieldName: 'Parse Trace Id From',
			placeholder: 'processor_trace_id_placeholder',
			name: ['trace_id', 'parse_from'],
			rules: [traceParserFieldValidator],
			dependencies: [
				['span_id', 'parse_from'],
				['trace_flags', 'parse_from'],
			],
		},
		{
			id: 3,
			fieldName: 'Parse Span Id From',
			placeholder: 'processor_span_id_placeholder',
			name: ['span_id', 'parse_from'],
			rules: [traceParserFieldValidator],
			dependencies: [
				['trace_id', 'parse_from'],
				['trace_flags', 'parse_from'],
			],
		},
		{
			id: 4,
			fieldName: 'Parse Trace flags From',
			placeholder: 'processor_trace_flags_placeholder',
			name: ['trace_flags', 'parse_from'],
			rules: [traceParserFieldValidator],
			dependencies: [
				['trace_id', 'parse_from'],
				['span_id', 'parse_from'],
			],
		},
	],
	time_parser: [
		{
			id: 1,
			fieldName: 'Name of Timestamp Parsing Processor',
			placeholder: 'processor_name_placeholder',
			name: 'name',
		},
		{
			id: 2,
			fieldName: 'Parse Timestamp Value From',
			placeholder: 'processor_parsefrom_placeholder',
			name: 'parse_from',
			initialValue: 'attributes.timestamp',
		},
		{
			id: 3,
			fieldName: 'Timestamp Format Type',
			placeholder: '',
			name: 'layout_type',
			initialValue: 'strptime',
			options: [
				{
					label: 'Unix Epoch',
					value: 'epoch',
				},
				{
					label: 'strptime Format',
					value: 'strptime',
				},
			],
			onFormValuesChanged: (
				changedValues: ProcessorData,
				form: FormInstance,
			): void => {
				if (changedValues?.layout_type) {
					const newLayoutValue =
						changedValues.layout_type === 'strptime' ? '%Y-%m-%dT%H:%M:%S.%f%z' : 's';

					form.setFieldValue('layout', newLayoutValue);
				}
			},
		},
		{
			id: 4,
			fieldName: 'Epoch Format',
			placeholder: '',
			name: 'layout',
			dependencies: ['layout_type'],
			shouldRender: (form: FormInstance): boolean => {
				const layoutType = form.getFieldValue('layout_type');
				return layoutType === 'epoch';
			},
			initialValue: 's',
			options: [
				{
					label: 'seconds',
					value: 's',
				},
				{
					label: 'milliseconds',
					value: 'ms',
				},
				{
					label: 'microseconds',
					value: 'us',
				},
				{
					label: 'nanoseconds',
					value: 'ns',
				},
				{
					label: 'seconds.milliseconds (eg: 1136214245.123)',
					value: 's.ms',
				},
				{
					label: 'seconds.microseconds (eg: 1136214245.123456)',
					value: 's.us',
				},
				{
					label: 'seconds.nanoseconds (eg: 1136214245.123456789)',
					value: 's.ns',
				},
			],
		},
		{
			id: 4,
			fieldName: 'Timestamp Format',
			placeholder: 'strptime directives based format. Eg: %Y-%m-%dT%H:%M:%S.%f%z',
			name: 'layout',
			dependencies: ['layout_type'],
			shouldRender: (form: FormInstance): boolean => {
				const layoutType = form.getFieldValue('layout_type');
				return layoutType === 'strptime';
			},
			initialValue: '%Y-%m-%dT%H:%M:%S.%f%z',
		},
	],
	severity_parser: [
		{
			id: 1,
			fieldName: 'Name of Severity Parsing Processor',
			placeholder: 'processor_name_placeholder',
			name: 'name',
		},
		{
			id: 2,
			fieldName: 'Parse Severity Value From',
			placeholder: 'processor_parsefrom_placeholder',
			name: 'parse_from',
			initialValue: 'attributes.logLevel',
		},
		{
			id: 3,
			fieldName: 'Values for level TRACE',
			placeholder: 'Specify comma separated values. Eg: trace, 0',
			name: ['mapping', 'trace'],
			rules: [],
			initialValue: ['trace'],
			compact: true,
		},
		{
			id: 4,
			fieldName: 'Values for level DEBUG',
			placeholder: 'Specify comma separated values. Eg: debug, 2xx',
			name: ['mapping', 'debug'],
			rules: [],
			initialValue: ['debug'],
			compact: true,
		},
		{
			id: 5,
			fieldName: 'Values for level INFO',
			placeholder: 'Specify comma separated values. Eg: info, 3xx',
			name: ['mapping', 'info'],
			rules: [],
			initialValue: ['info'],
			compact: true,
		},
		{
			id: 6,
			fieldName: 'Values for level WARN',
			placeholder: 'Specify comma separated values. Eg: warning, 4xx',
			name: ['mapping', 'warn'],
			rules: [],
			initialValue: ['warn'],
			compact: true,
		},
		{
			id: 7,
			fieldName: 'Values for level ERROR',
			placeholder: 'Specify comma separated values. Eg: error, 5xx',
			name: ['mapping', 'error'],
			rules: [],
			initialValue: ['error'],
			compact: true,
		},
		{
			id: 8,
			fieldName: 'Values for level FATAL',
			placeholder: 'Specify comma separated values. Eg: fatal, panic',
			name: ['mapping', 'fatal'],
			rules: [],
			initialValue: ['fatal'],
			compact: true,
		},
		{
			id: 9,
			fieldName: 'Override Severity Text',
			placeholder:
				'Should the parsed severity set both severity and severityText?',
			name: ['overwrite_text'],
			rules: [],
			initialValue: true,
			hidden: true,
		},
	],
	retain: [
		{
			id: 1,
			fieldName: 'Name of Retain Processor',
			placeholder: 'processor_name_placeholder',
			name: 'name',
		},
		{
			id: 2,
			fieldName: 'Fields',
			placeholder: 'processor_fields_placeholder',
			name: 'fields',
		},
	],
	move: [
		{
			id: 1,
			fieldName: 'Name of Move Processor',
			placeholder: 'processor_name_placeholder',
			name: 'name',
		},
		{
			id: 2,
			fieldName: 'From',
			placeholder: 'processor_from_placeholder',
			name: 'from',
		},
		{
			id: 3,
			fieldName: 'To',
			placeholder: 'processor_to_placeholder',
			name: 'to',
		},
	],
	copy: [
		{
			id: 1,
			fieldName: 'Name of Copy Processor',
			placeholder: 'processor_name_placeholder',
			name: 'name',
		},
		{
			id: 2,
			fieldName: 'From',
			placeholder: 'processor_from_placeholder',
			name: 'from',
		},
		{
			id: 3,
			fieldName: 'To',
			placeholder: 'processor_to_placeholder',
			name: 'to',
		},
	],
};

export const PREDEFINED_MAPPING = {
	environment: ['service.env', 'environment', 'env'],
	host: ['host', 'hostname', 'host.name'],
	message: ['message', 'msg', 'log'],
	service: ['service', 'appname'],
	severity: ['status', 'severity', 'level'],
	span_id: ['span_id', 'span.id'],
	trace_flags: ['flags'],
	trace_id: ['trace_id', 'trace.id'],
};
