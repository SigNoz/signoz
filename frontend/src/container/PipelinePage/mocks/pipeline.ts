import { Pipeline, PipelineData } from 'types/api/pipeline/def';
import { DataTypes } from 'types/api/queryBuilder/queryAutocompleteResponse';
import { TagFilter } from 'types/api/queryBuilder/queryBuilderData';

export const configurationVersion = '1.0';

export function mockPipelineFilter(
	key: string,
	op: string,
	value: string,
): TagFilter {
	return {
		op: 'AND',
		items: [
			{
				id: `${key}-${value}`,
				key: {
					key,
					dataType: DataTypes.String,
					type: '',
					isColumn: false,
					isJSON: false,
				},
				op,
				value,
			},
		],
	};
}

export const pipelineMockData: Array<PipelineData> = [
	{
		id: '4453c8b0-c0fd-42bf-bf09-7cc1b04ccdc9',
		orderId: 1,
		name: 'Apache common parser',
		alias: 'apachecommonparser',
		description: 'This is a desc',
		enabled: false,
		filter: mockPipelineFilter('source', '=', 'nginx'),
		config: [
			{
				orderId: 1,
				enabled: true,
				type: 'grok_parser',
				id: 'grokusecommon',
				name: 'grok use common asd',
				output: 'renameauth',
				parse_to: 'attributes',
				pattern: '%{COMMONAPACHELOG}',
				parse_from: 'body',
			},
			{
				orderId: 2,
				enabled: true,
				type: 'move',
				id: 'renameauth',
				name: 'rename auth',
				from: 'attributes.auth',
				to: 'attributes.username',
			},
			{
				orderId: 3,
				enabled: true,
				type: 'json_parser',
				id: 'jsonparser',
				name: 'json parser',
				from: 'attributes.auth',
				to: 'attributes.username',
			},
		],
		createdBy: 'nityananda@signoz.io',
		createdAt: '2023-03-07T16:56:53.36071141Z',
	},
	{
		id: 'a3675a0c-ff73-4ddb-be39-4351ace69231',
		orderId: 2,
		name: 'Moving pipeline new',
		alias: 'movingpipelinenew',
		description: 'This is a desc of move',
		enabled: false,
		filter: mockPipelineFilter('method', '=', 'POST'),
		config: [
			{
				orderId: 1,
				enabled: true,
				type: 'copy',
				id: 'mv1',
				name: 'mymove',
				from: 'attributes.method',
				to: 'attributes.method11',
			},
		],
		createdBy: 'chintan@signoz.io',
		createdAt: '2023-03-07T16:55:27.789595116Z',
	},
];

export const pipelineApiResponseMockData: Pipeline = {
	id: '67ace08a-6b6c-4221-ab58-a5d3bd5eb6f2',
	version: 5,
	elementType: 'log_pipelines',
	createdBy: 'test@signoz.io',
	active: false,
	is_valid: false,
	disabled: false,
	deployStatus: 'IN_PROGRESS',
	deployResult: 'Deployment started',
	lastHash: 'q��҂�&覣ʝup��\u0003���q�6�\u001e�<ѥIb�',
	lastConf:
		'[{"id":"a3675a0c-ff73-4ddb-be39-4351ace69231","orderId":"1","name":"Moving pipeline new","alias":"movingpipelinenew","enabled":false,"filter":"attributes.method == POST","config":[{"type":"copy","id":"mv1","name":"mymove","from":"attributes.method","to":"attributes.method11"}],"createdBy":"nityananda@signoz.io","createdAt":"2023-03-07T16:55:27.789595116Z"},{"id":"4453c8b0-c0fd-42bf-bf09-7cc1b04ccdc9","orderId":"2","name":"Apache common parser","alias":"apachecommonparser","enabled":false,"filter":"attributes.source == nginx","config":[{"type":"grok_parser","id":"grokusecommon","name":"grok use common asd","output":"renameauth","parse_to":"attributes","pattern":"%{COMMONAPACHELOG}","parse_from":"body"},{"type":"move","id":"renameauth","name":"rename auth","from":"attributes.auth","to":"attributes.username"}],"createdBy":"nityananda@signoz.io","createdAt":"2023-03-07T16:56:53.36071141Z"}]',
	pipelines: pipelineMockData,
	history: [
		{
			id: 'e118dedd-e996-455a-9cb2-5bf50b77fc35',
			version: 1,
			elementType: 'log_pipelines',
			createdBy: 'test2@signoz.io',
			active: false,
			isValid: false,
			disabled: false,
			deployStatus: 'DEPLOYED',
			deployResult: 'deploy successful',
			lastHash: '',
			lastConf: '',
			// eslint-disable-next-line sonarjs/no-duplicate-string
			createdAt: '2021-03-07T16:56:53.36071141Z',
			createdByName: 'test2',
		},
		{
			id: '9a98673d-b5db-4281-89d3-d85ed9ffe311',
			version: 2,
			elementType: 'log_pipelines',
			createdBy: 'test3@signoz.io',
			active: false,
			isValid: false,
			disabled: false,
			deployStatus: 'DEPLOYED',
			deployResult: 'deploy successful',
			lastHash: '',
			lastConf: '',
			createdAt: '2021-03-07T16:56:53.36071141Z',
			createdByName: 'test3',
		},
		{
			id: '9fdb0813-f77f-4837-815e-bb6eedd64f68',
			version: 3,
			elementType: 'log_pipelines',
			createdBy: 'nityananda+1@signoz.io',
			active: false,
			isValid: false,
			disabled: false,
			deployStatus: 'IN_PROGRESS',
			deployResult: '',
			lastHash: '',
			lastConf: '',
			createdAt: '2021-03-07T16:56:53.36071141Z',
			createdByName: 'nityananda+1',
		},
		{
			id: '87efb1cf-85b0-4aa4-934e-62a118fa4ec7',
			version: 4,
			elementType: 'log_pipelines',
			createdBy: 'nityananda+2@signoz.io',
			active: false,
			isValid: false,
			disabled: false,
			deployStatus: 'IN_PROGRESS',
			deployResult: '',
			lastHash: '',
			lastConf: '',
			createdAt: '2021-03-07T16:56:53.36071141Z',
			createdByName: 'nityananda+2',
		},
		{
			id: '67ace08a-6b6c-4221-ab58-a5d3bd5eb6f2',
			version: 5,
			elementType: 'log_pipelines',
			createdBy: 'nityananda+4@signoz.io',
			active: false,
			isValid: false,
			disabled: false,
			deployStatus: 'IN_PROGRESS',
			deployResult: '',
			lastHash: '',
			lastConf: '',
			createdAt: '2021-03-07T16:56:53.36071141Z',
			createdByName: 'nityananda+4',
		},
	],
};
