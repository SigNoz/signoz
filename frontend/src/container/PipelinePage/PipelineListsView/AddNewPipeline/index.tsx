import { Button, Divider, Form, Modal } from 'antd';
import { useAppContext } from 'providers/App/App';
import React, { useCallback, useEffect, useMemo } from 'react';
import { useTranslation } from 'react-i18next';
import { ActionMode, ActionType, PipelineData } from 'types/api/pipeline/def';
import { v4 } from 'uuid';

import { ModalButtonWrapper, ModalTitle } from '../styles';
import { getEditedDataSource, getRecordIndex } from '../utils';
import { renderPipelineForm } from './utils';

function AddNewPipeline({
	isActionType,
	setActionType,
	selectedPipelineData,
	setShowSaveButton,
	setCurrPipelineData,
	currPipelineData,
}: AddNewPipelineProps): JSX.Element {
	const [form] = Form.useForm();
	const { t } = useTranslation('pipeline');
	const { user } = useAppContext();

	const isEdit = isActionType === 'edit-pipeline';
	const isAdd = isActionType === 'add-pipeline';

	useEffect(() => {
		if (isEdit) {
			form.setFieldsValue(selectedPipelineData);
		}
		if (isAdd) {
			form.resetFields();
		}
	}, [form, isEdit, isAdd, selectedPipelineData]);

	const onFinish = (values: PipelineData): void => {
		const newPipeLineData: PipelineData = {
			id: v4(),
			orderId: (currPipelineData?.length || 0) + 1,
			createdAt: new Date().toISOString(),
			createdBy: user?.name || '',
			name: values.name,
			alias: values.name.replace(/\s/g, ''),
			description: values.description,
			filter: values.filter,
			config: [],
			enabled: true,
		};

		if (isEdit && selectedPipelineData) {
			const findRecordIndex = getRecordIndex(
				currPipelineData,
				selectedPipelineData,
				'id',
			);
			const updatedPipelineData: PipelineData = {
				...currPipelineData[findRecordIndex],
				...values,
			};

			const editedPipelineData = getEditedDataSource(
				currPipelineData,
				selectedPipelineData,
				'id',
				updatedPipelineData,
			);

			setCurrPipelineData(editedPipelineData);
		}
		if (isAdd) {
			setCurrPipelineData((prevState) => {
				if (prevState) return [...prevState, newPipeLineData];
				return [newPipeLineData];
			});
		}
		setActionType(undefined);
	};

	const onCancelModalHandler = (): void => {
		setActionType(undefined);
	};

	const modalTitle = useMemo(
		(): string =>
			isEdit
				? `${t('edit_pipeline')} : ${selectedPipelineData?.name}`
				: t('create_pipeline'),
		[isEdit, selectedPipelineData?.name, t],
	);

	const onOkModalHandler = useCallback(
		() => setShowSaveButton(ActionMode.Editing),
		[setShowSaveButton],
	);

	const isOpen = useMemo(() => isEdit || isAdd, [isAdd, isEdit]);

	return (
		<Modal
			title={<ModalTitle level={4}>{modalTitle}</ModalTitle>}
			centered
			open={isOpen}
			width={800}
			footer={null}
			onCancel={onCancelModalHandler}
		>
			<Divider plain />
			<Form
				name="add-new-pipeline"
				layout="vertical"
				onFinish={onFinish}
				autoComplete="off"
				form={form}
			>
				{renderPipelineForm()}
				<Divider plain />
				<Form.Item>
					<ModalButtonWrapper>
						<Button
							key="submit"
							type="primary"
							htmlType="submit"
							onClick={onOkModalHandler}
						>
							{isEdit ? t('update') : t('create')}
						</Button>
						<Button key="cancel" onClick={onCancelModalHandler}>
							{t('cancel')}
						</Button>
					</ModalButtonWrapper>
				</Form.Item>
			</Form>
		</Modal>
	);
}

interface AddNewPipelineProps {
	isActionType: string;
	setActionType: (actionType?: ActionType) => void;
	selectedPipelineData: PipelineData | undefined;
	setShowSaveButton: (actionMode: ActionMode) => void;
	setCurrPipelineData: (
		value: React.SetStateAction<Array<PipelineData>>,
	) => void;
	currPipelineData: Array<PipelineData>;
}

export default AddNewPipeline;
