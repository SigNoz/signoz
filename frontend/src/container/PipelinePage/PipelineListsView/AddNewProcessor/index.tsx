import { Button, Divider, Form, Modal } from 'antd';
import { useCallback, useEffect, useMemo, useState } from 'react';
import { useTranslation } from 'react-i18next';
import {
	ActionMode,
	ActionType,
	PipelineData,
	ProcessorData,
} from 'types/api/pipeline/def';
import { v4 } from 'uuid';

import { ModalButtonWrapper, ModalTitle } from '../styles';
import { getEditedDataSource, getRecordIndex } from '../utils';
import { DEFAULT_PROCESSOR_TYPE, processorFields } from './config';
import TypeSelect from './FormFields/TypeSelect';
import ProcessorForm from './ProcessorForm';

function AddNewProcessor({
	isActionType,
	setActionType,
	selectedProcessorData,
	setShowSaveButton,
	expandedPipelineData,
	setExpandedPipelineData,
}: AddNewProcessorProps): JSX.Element {
	const [form] = Form.useForm();
	const { t } = useTranslation('pipeline');
	const [processorType, setProcessorType] = useState<string>(
		DEFAULT_PROCESSOR_TYPE,
	);

	const isEdit = isActionType === 'edit-processor';
	const isAdd = isActionType === 'add-processor';

	useEffect(() => {
		if (isEdit || isAdd) {
			// Reset form first to clear any stale fields from previous processors
			form.resetFields();
		}
		if (isEdit && selectedProcessorData && expandedPipelineData?.config) {
			const findRecordIndex = getRecordIndex(
				expandedPipelineData?.config,
				selectedProcessorData,
				'id',
			);

			const updatedProcessorData = {
				...expandedPipelineData?.config?.[findRecordIndex],
			};
			setProcessorType(updatedProcessorData.type);
			form.setFieldsValue(updatedProcessorData);
		}
	}, [form, isEdit, isAdd, selectedProcessorData, expandedPipelineData?.config]);

	const handleProcessorType = (value: string | unknown): void => {
		const typedValue = String(value) || DEFAULT_PROCESSOR_TYPE;
		setProcessorType(typedValue);
	};

	const onFinish = (values: { name: string }): void => {
		const totalDataLength = expandedPipelineData?.config?.length || 0;

		const newProcessorData = {
			id: v4(),
			orderId: Number(totalDataLength || 0) + 1,
			type: processorType,
			enabled: true,
			...values,
		};

		if (isEdit && selectedProcessorData && expandedPipelineData?.config) {
			const findRecordIndex = getRecordIndex(
				expandedPipelineData?.config,
				selectedProcessorData,
				'id',
			);

			const processorData = expandedPipelineData?.config?.[findRecordIndex];

			const updatedProcessorData = {
				id: processorData?.id || v4(),
				orderId: processorData?.orderId,
				type: processorType,
				enabled: processorData?.enabled,
				output: processorData?.output,
				...values,
			};

			const editedData = getEditedDataSource(
				expandedPipelineData.config,
				selectedProcessorData,
				'name',
				updatedProcessorData,
			);

			const modifiedProcessorData = { ...expandedPipelineData };

			modifiedProcessorData.config = editedData;

			setExpandedPipelineData(modifiedProcessorData);
		}
		if (isAdd && expandedPipelineData) {
			const modifiedProcessorData = {
				...expandedPipelineData,
			};
			if (
				modifiedProcessorData.config !== undefined &&
				modifiedProcessorData.config
			) {
				modifiedProcessorData.config = [
					...modifiedProcessorData.config,
					newProcessorData,
				];
				if (totalDataLength > 0) {
					modifiedProcessorData.config[totalDataLength - 1].output =
						newProcessorData.id;
				}
			}
			setExpandedPipelineData(modifiedProcessorData);
		}
		setActionType(undefined);
		handleProcessorType(DEFAULT_PROCESSOR_TYPE);
	};

	const onCancelModal = (): void => {
		setActionType(undefined);
		handleProcessorType(DEFAULT_PROCESSOR_TYPE);
	};

	const modalTitle = useMemo(
		(): string =>
			isEdit
				? `${t('edit_processor')} ${selectedProcessorData?.name}`
				: t('create_processor'),
		[isEdit, selectedProcessorData?.name, t],
	);

	const onOkModalHandler = useCallback(
		() => setShowSaveButton(ActionMode.Editing),
		[setShowSaveButton],
	);

	const isOpen = useMemo(() => isEdit || isAdd, [isAdd, isEdit]);

	const onFormValuesChanged = useCallback(
		(changedValues: ProcessorData): void => {
			processorFields[processorType].forEach((field) => {
				if (field.onFormValuesChanged) {
					field.onFormValuesChanged(changedValues, form);
				}
			});
		},
		[form, processorType],
	);

	return (
		<Modal
			title={<ModalTitle level={4}>{modalTitle}</ModalTitle>}
			centered
			open={isOpen}
			width={800}
			footer={null}
			onCancel={onCancelModal}
			destroyOnClose
		>
			<Divider plain />
			<Form
				name="add-new-processor"
				layout="vertical"
				onFinish={onFinish}
				autoComplete="off"
				form={form}
				onValuesChange={onFormValuesChanged}
			>
				<TypeSelect value={processorType} onChange={handleProcessorType} />
				<ProcessorForm
					processorType={processorType}
					selectedProcessorData={selectedProcessorData}
					isAdd={isAdd}
				/>
				<Divider plain />
				<Form.Item>
					<ModalButtonWrapper>
						<Button
							key="submit"
							type="primary"
							htmlType="submit"
							onClick={onOkModalHandler}
						>
							{isEdit ? t('update') : t('create')}
						</Button>
						<Button key="cancel" onClick={onCancelModal}>
							{t('cancel')}
						</Button>
					</ModalButtonWrapper>
				</Form.Item>
			</Form>
		</Modal>
	);
}

AddNewProcessor.defaultProps = {
	selectedProcessorData: undefined,
	expandedPipelineData: {},
};

interface AddNewProcessorProps {
	isActionType: string;
	setActionType: (actionType?: ActionType) => void;
	selectedProcessorData?: ProcessorData;
	setShowSaveButton: (actionMode: ActionMode) => void;
	expandedPipelineData?: PipelineData;
	setExpandedPipelineData: (data: PipelineData) => void;
}

export default AddNewProcessor;
