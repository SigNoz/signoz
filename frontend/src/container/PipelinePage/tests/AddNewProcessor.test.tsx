import { render } from '@testing-library/react';
import { I18nextProvider } from 'react-i18next';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import i18n from 'ReactI18';
import store from 'store';

import { pipelineMockData } from '../mocks/pipeline';
import AddNewProcessor from '../PipelineListsView/AddNewProcessor';
import { matchMedia } from './AddNewPipeline.test';

jest.mock('uplot', () => {
	const paths = {
		spline: jest.fn(),
		bars: jest.fn(),
	};
	const uplotMock = jest.fn(() => ({
		paths,
	}));
	return {
		paths,
		default: uplotMock,
	};
});

beforeAll(() => {
	matchMedia();
});

const selectedProcessorData = {
	id: '1',
	orderId: 1,
	type: 'grok_parser',
	name: 'grok use common',
	output: 'grokusecommon',
};
describe('PipelinePage container test', () => {
	it('should render AddNewProcessor section', () => {
		const setActionType = jest.fn();
		const isActionType = 'add-processor';

		const { asFragment } = render(
			<MemoryRouter>
				<Provider store={store}>
					<I18nextProvider i18n={i18n}>
						<AddNewProcessor
							isActionType={isActionType}
							setActionType={setActionType}
							selectedProcessorData={selectedProcessorData}
							setShowSaveButton={jest.fn()}
							expandedPipelineData={pipelineMockData[0]}
							setExpandedPipelineData={jest.fn()}
						/>
					</I18nextProvider>
				</Provider>
			</MemoryRouter>,
		);
		expect(asFragment()).toMatchSnapshot();
	});
});
