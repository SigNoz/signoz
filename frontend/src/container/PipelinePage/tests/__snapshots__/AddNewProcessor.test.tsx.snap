// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PipelinePage container test should render AddNewProcessor section 1`] = `<DocumentFragment />`;
